## Habitable Planet Data Cleaning

In [129]:
import pandas as pd
import numpy as np

In [130]:
k_data = pd.read_csv("planet.csv")
k_data.columns = ['planet_name', 'num_stars', 'controv_flag', 'orbital_period', 'planet_semi-major_axis', 'planet_radius', 'planet_mass', 'eccentricity', 'insolation_flux', 'equi_temp', 'spectral_type', 'stellar_temp', 'stellar_radius', 'stellar_mass', 'distance']

planet_name = Name of Planet

num_stars = Number of stars

controv_flag = 0 if no controversy, else 1

orbital_period = orbital period [days]

planet_semi-major_axis = distance from star [au]

planet_radius = planet radius [Earth Radius]

planet_mass = planet mass [Earth Mass]

eccentricity = amount by which orbit of planet deviates from a perfect circle (instability in temp)

insolation_flux = Another form of equi_temp [Earth]

equi_temp = planetary equilibrium (The equilibrium temperature of the planet as modeled by a black body heated only by its host star, or for directly imaged planets, the effective temperature of the planet required to match the measured luminosity if the planet were a black body)

spectral_type = Classification based upon Morgan-Keenan system. (Hottest) O, B, A, F, G, K, and M (Coolest). Each letter class is then subdivided using a numeric digit with 0 being hottest and 9 being coolest. Luminosity class 0 or Ia+ is used for hypergiants, class I for supergiants, class II for bright giants, class III for regular giants, class IV for subgiants, class V for main-sequence stars, class sd (or VI) for subdwarfs, and class D (or VII) for white dwarfs.

stellar_radius = radius of star [Sol Radius]

stellar_mass = mass of star [Sol Mass]

distance = distance from Earth [parsec]

In [131]:
k_data.head()

planet_name  num_stars  controv_flag  orbital_period  \
0    11 Com b          2             0             NaN   
1    11 Com b          2             0       326.03000   
2    11 Com b          2             0       323.21000   
3    11 UMi b          1             0       516.21997   
4    11 UMi b          1             0             NaN   

   planet_semi-major_axis  planet_radius  planet_mass  eccentricity  \
0                   1.210            NaN   5434.70000           NaN   
1                   1.290            NaN   6165.60000         0.231   
2                   1.178            NaN   4914.89849         0.238   
3                   1.530            NaN   4684.81420         0.080   
4                   1.510            NaN   3432.40000           NaN   

   insolation_flux  equi_temp spectral_type  stellar_temp  stellar_radius  \
0              NaN        NaN           NaN           NaN             NaN   
1              NaN        NaN        G8 III        4742.0           19.00   
2              NaN        NaN        G8 III        4874.0           13.76   
3              NaN        NaN           NaN        4213.0           29.79   
4              NaN        NaN           NaN           NaN             NaN   

   stellar_mass  distance  
0          2.60   93.1846  
1          2.70   93.1846  
2          2.09   93.1846  
3          2.78  125.3210  
4          1.70  125.3210

In [132]:
#Total rows
k_data['planet_name'].count()

35896

In [133]:
#Total planets
k_data['planet_name'].nunique()

5602

In [134]:
#Take duplicates and average the data
avg_data = k_data.groupby('planet_name').mean(numeric_only=True).reset_index()
avg_data.reset_index(drop=True, inplace=True)
avg_data.tail()

planet_name  num_stars  controv_flag  orbital_period  \
5597   ups And b        2.0           0.0        4.616229   
5598   ups And c        2.0           0.0      240.728533   
5599   ups And d        2.0           0.0     1285.346167   
5600   ups Leo b        1.0           0.0      385.200000   
5601    xi Aql b        1.0           0.0      136.860000   

      planet_semi-major_axis  planet_radius  planet_mass  eccentricity  \
5597                0.058895            NaN   219.960419      0.030200   
5598                0.827712            NaN  1275.328758      0.238933   
5599                2.528382            NaN  1616.242590      0.281117   
5600                1.180000            NaN   162.092490      0.320000   
5601                0.628333            NaN   716.301647      0.029500   

      insolation_flux  equi_temp  stellar_temp  stellar_radius  stellar_mass  \
5597              NaN        NaN   6137.820000           1.615        1.2500   
5598              NaN        NaN   6117.093333           1.615        1.2625   
5599              NaN        NaN   6117.093333           1.615        1.2625   
5600              NaN        NaN   4836.000000          11.220        1.4800   
5601              NaN        NaN   4810.500000          11.430        1.7800   

      distance  
5597   13.4054  
5598   13.4054  
5599   13.4054  
5600   52.5973  
5601   56.1858

In [135]:
def first_non_zero(s):
    non_zero_values = s[s != '0']
    if len(non_zero_values) > 0:
        return non_zero_values.iloc[0]
    else:
        return '0'

In [136]:
# Save Spectral_Type data
k_data['spectral_type'] = k_data['spectral_type'].fillna('0').astype(str)
spectral = k_data.groupby('planet_name')['spectral_type'].agg(first_non_zero)
avg_data_reset_index = avg_data.reset_index()
avg_data_with_spectral = pd.merge(avg_data_reset_index, spectral, on='planet_name', how='left')
avg_data_with_spectral

index planet_name  num_stars  controv_flag  orbital_period  \
0         0    11 Com b        2.0           0.0      324.620000   
1         1    11 UMi b        1.0           0.0      516.219985   
2         2    14 And b        1.0           0.0      186.300000   
3         3    14 Her b        1.0           0.0     1766.378417   
4         4  16 Cyg B b        3.0           0.0      799.375000   
...     ...         ...        ...           ...             ...   
5597   5597   ups And b        2.0           0.0        4.616229   
5598   5598   ups And c        2.0           0.0      240.728533   
5599   5599   ups And d        2.0           0.0     1285.346167   
5600   5600   ups Leo b        1.0           0.0      385.200000   
5601   5601    xi Aql b        1.0           0.0      136.860000   

      planet_semi-major_axis  planet_radius  planet_mass  eccentricity  \
0                   1.226000            NaN  5505.066163      0.234500   
1                   1.526667            NaN  3818.094733      0.080000   
2                   0.761667            NaN  1224.550433      0.000000   
3                   2.814750            NaN  1642.383591      0.362925   
4                   1.662833            NaN   533.514528      0.676033   
...                      ...            ...          ...           ...   
5597                0.058895            NaN   219.960419      0.030200   
5598                0.827712            NaN  1275.328758      0.238933   
5599                2.528382            NaN  1616.242590      0.281117   
5600                1.180000            NaN   162.092490      0.320000   
5601                0.628333            NaN   716.301647      0.029500   

      insolation_flux  equi_temp  stellar_temp  stellar_radius  stellar_mass  \
0                 NaN        NaN   4808.000000       16.380000      2.463333   
1                 NaN        NaN   4276.500000       26.935000      2.093333   
2                 NaN        NaN   4850.500000       11.275000      1.726667   
3                 NaN        NaN   5296.985000        0.976667      0.927143   
4                 NaN        NaN   5728.594000        1.140000      1.016000   
...               ...        ...           ...             ...           ...   
5597              NaN        NaN   6137.820000        1.615000      1.250000   
5598              NaN        NaN   6117.093333        1.615000      1.262500   
5599              NaN        NaN   6117.093333        1.615000      1.262500   
5600              NaN        NaN   4836.000000       11.220000      1.480000   
5601              NaN        NaN   4810.500000       11.430000      1.780000   

      distance spectral_type  
0      93.1846        G8 III  
1     125.3210        K4 III  
2      75.4392        K0 III  
3      17.9323          K0 V  
4      21.1397        G2.5 V  
...        ...           ...  
5597   13.4054          F8 V  
5598   13.4054          F8 V  
5599   13.4054          F8 V  
5600   52.5973        G9 III  
5601   56.1858        K0 III  

[5602 rows x 16 columns]

In [137]:
avg_data_with_spectral.rename(columns={'spectral_type_y' : 'spectral_type'}, inplace=True)
avg_data = avg_data_with_spectral
#avg_data.drop(columns=['spectral_type_x'], inplace=True)
avg_data.head()

index planet_name  num_stars  controv_flag  orbital_period  \
0      0    11 Com b        2.0           0.0      324.620000   
1      1    11 UMi b        1.0           0.0      516.219985   
2      2    14 And b        1.0           0.0      186.300000   
3      3    14 Her b        1.0           0.0     1766.378417   
4      4  16 Cyg B b        3.0           0.0      799.375000   

   planet_semi-major_axis  planet_radius  planet_mass  eccentricity  \
0                1.226000            NaN  5505.066163      0.234500   
1                1.526667            NaN  3818.094733      0.080000   
2                0.761667            NaN  1224.550433      0.000000   
3                2.814750            NaN  1642.383591      0.362925   
4                1.662833            NaN   533.514528      0.676033   

   insolation_flux  equi_temp  stellar_temp  stellar_radius  stellar_mass  \
0              NaN        NaN      4808.000       16.380000      2.463333   
1              NaN        NaN      4276.500       26.935000      2.093333   
2              NaN        NaN      4850.500       11.275000      1.726667   
3              NaN        NaN      5296.985        0.976667      0.927143   
4              NaN        NaN      5728.594        1.140000      1.016000   

   distance spectral_type  
0   93.1846        G8 III  
1  125.3210        K4 III  
2   75.4392        K0 III  
3   17.9323          K0 V  
4   21.1397        G2.5 V

In [138]:
# Checking NAN counts for each row.
for column in avg_data.columns:
    nan_count = avg_data[column].isna().sum()
    print(f"Number of NaN values in '{column}': {nan_count}")

Number of NaN values in 'index': 0
Number of NaN values in 'planet_name': 0
Number of NaN values in 'num_stars': 0
Number of NaN values in 'controv_flag': 0
Number of NaN values in 'orbital_period': 249
Number of NaN values in 'planet_semi-major_axis': 296
Number of NaN values in 'planet_radius': 1361
Number of NaN values in 'planet_mass': 2907
Number of NaN values in 'eccentricity': 777
Number of NaN values in 'insolation_flux': 1710
Number of NaN values in 'equi_temp': 1462
Number of NaN values in 'stellar_temp': 392
Number of NaN values in 'stellar_radius': 535
Number of NaN values in 'stellar_mass': 28
Number of NaN values in 'distance': 118
Number of NaN values in 'spectral_type': 0


In [139]:
#Remove rows that have controversy
avg_data_filtered = avg_data.loc[avg_data['controv_flag'] != 1]
avg_data_filtered.reset_index(drop=True, inplace=True)
avg_data_filtered['controv_flag'].nunique()

1

In [140]:
proj_data = avg_data_filtered
proj_data.head()

index planet_name  num_stars  controv_flag  orbital_period  \
0      0    11 Com b        2.0           0.0      324.620000   
1      1    11 UMi b        1.0           0.0      516.219985   
2      2    14 And b        1.0           0.0      186.300000   
3      3    14 Her b        1.0           0.0     1766.378417   
4      4  16 Cyg B b        3.0           0.0      799.375000   

   planet_semi-major_axis  planet_radius  planet_mass  eccentricity  \
0                1.226000            NaN  5505.066163      0.234500   
1                1.526667            NaN  3818.094733      0.080000   
2                0.761667            NaN  1224.550433      0.000000   
3                2.814750            NaN  1642.383591      0.362925   
4                1.662833            NaN   533.514528      0.676033   

   insolation_flux  equi_temp  stellar_temp  stellar_radius  stellar_mass  \
0              NaN        NaN      4808.000       16.380000      2.463333   
1              NaN        NaN      4276.500       26.935000      2.093333   
2              NaN        NaN      4850.500       11.275000      1.726667   
3              NaN        NaN      5296.985        0.976667      0.927143   
4              NaN        NaN      5728.594        1.140000      1.016000   

   distance spectral_type  
0   93.1846        G8 III  
1  125.3210        K4 III  
2   75.4392        K0 III  
3   17.9323          K0 V  
4   21.1397        G2.5 V

In [141]:
proj_data2 = proj_data.drop(columns=['controv_flag'])
proj_data2.head()

index planet_name  num_stars  orbital_period  planet_semi-major_axis  \
0      0    11 Com b        2.0      324.620000                1.226000   
1      1    11 UMi b        1.0      516.219985                1.526667   
2      2    14 And b        1.0      186.300000                0.761667   
3      3    14 Her b        1.0     1766.378417                2.814750   
4      4  16 Cyg B b        3.0      799.375000                1.662833   

   planet_radius  planet_mass  eccentricity  insolation_flux  equi_temp  \
0            NaN  5505.066163      0.234500              NaN        NaN   
1            NaN  3818.094733      0.080000              NaN        NaN   
2            NaN  1224.550433      0.000000              NaN        NaN   
3            NaN  1642.383591      0.362925              NaN        NaN   
4            NaN   533.514528      0.676033              NaN        NaN   

   stellar_temp  stellar_radius  stellar_mass  distance spectral_type  
0      4808.000       16.380000      2.463333   93.1846        G8 III  
1      4276.500       26.935000      2.093333  125.3210        K4 III  
2      4850.500       11.275000      1.726667   75.4392        K0 III  
3      5296.985        0.976667      0.927143   17.9323          K0 V  
4      5728.594        1.140000      1.016000   21.1397        G2.5 V

# Factoring in Earth

Mass = 5.9722×10^24 kg

Eccentricity =0.0034 and up to 0.058. AVG = 0.0307

Radius = 3,958.8 miles

Orbital Period = 365

Semi-major Axis of Orbit = 149,598,023 km (or 92,955,902 miles)

Insolation Flux = 1400 W/m

Equilibrium Temp = 255 K

Spectral Type = G-type (G2-V)

Solar Mass = 1.989 × 10^30 kg

Sol Radius = 695,700 kilometres

In [142]:
columns = ['planet_name', 'num_stars', 'orbital_period', 'planet_semi-major_axis', 'planet_radius', 'planet_mass', 'eccentricity', 'insolation_flux', 'equi_temp', 'stellar_temp', 'stellar_radius', 'stellar_mass', 'distance']
earth = pd.DataFrame(columns=columns)
earth_data = {'planet_name' : 'Earth',
              'num_stars' : 1,
              'orbital_period' : 365,
              'planet_semi-major_axis' : 1,
              'planet_radius' : 1,
              'planet_mass' : 1,
              'eccentricity' : 0.0307,
              'insolation_flux' : 0,
              'equi_temp' : 255,
              'stellar_temp' : 5800,
              'stellar_radius' : 1,
              'stellar_mass' : 1,
              'distance' : 0
}
earth = earth.append(earth_data, ignore_index=True)
earth

/tmp/ipykernel_130/224570154.py:17: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  earth = earth.append(earth_data, ignore_index=True)


planet_name num_stars orbital_period planet_semi-major_axis planet_radius  \
0       Earth         1            365                      1             1   

  planet_mass  eccentricity insolation_flux equi_temp stellar_temp  \
0           1        0.0307               0       255         5800   

  stellar_radius stellar_mass distance  
0              1            1        0

# Steps

1) Calculate planets in habitable zone.

    a) Planetary Equilibrium Temp is easiest. 175 K - 270 K
    
    b) Semi-major axis (distance from spectral body) and spectral type / stellar temp to estimate equi_temp
    


2) Clean data with planets that do not reside in the habitable zone

In [143]:
# How many planets are habitable based upon our current count?
k_data = proj_data2
habitable = k_data[(k_data['equi_temp'] >= 175) & (k_data['equi_temp'] <= 270)]
count = habitable['planet_name'].count()
missing = k_data['equi_temp'].isna().sum()
print(count, "planets with equilibrium temp within acceptable habitable zone parameters.", missing, "NAN values")

78 planets with equilibrium temp within acceptable habitable zone parameters. 1442 NAN values


In [144]:
habitable

index         planet_name  num_stars  orbital_period  \
171     172    EPIC 248847494 b        1.0     3650.000000   
179     180           GJ 1002 b        1.0       10.346500   
180     181           GJ 1002 c        1.0       21.202000   
237     239            GJ 357 d        1.0       55.661000   
263     265            GJ 514 b        1.0      140.430000   
...     ...                 ...        ...             ...   
5292   5318        TRAPPIST-1 g        1.0       12.352981   
5304   5330  Teegarden's Star c        1.0       11.412500   
5433   5461           WASP-41 c        1.0      421.000000   
5440   5468           WASP-47 c        1.0      582.500000   
5512   5540         Wolf 1069 b        1.0       15.564000   

      planet_semi-major_axis  planet_radius  planet_mass  eccentricity  \
171                 4.500000      12.442000  4131.790000       0.00000   
179                 0.045700            NaN     1.080000           NaN   
180                 0.073800            NaN     1.360000           NaN   
237                 0.204000            NaN     6.100000           NaN   
263                 0.422000            NaN     5.200000       0.45000   
...                      ...            ...          ...           ...   
5292                0.046406       1.134667     1.269667       0.03154   
5304                0.044900            NaN     1.080000       0.02000   
5433                1.070000            NaN  1010.699400       0.29400   
5440                1.387667            NaN   422.802300       0.27025   
5512                0.067200            NaN     1.260000           NaN   

      insolation_flux  equi_temp  stellar_temp  stellar_radius  stellar_mass  \
171          0.190000      183.0   4898.000000          2.7000        0.9000   
179          0.670000      231.0   3024.000000          0.1400        0.1200   
180          0.260000      182.0   3024.000000          0.1400        0.1200   
237          0.380000      220.0   3505.000000          0.3400        0.3400   
263          0.280000      202.0   3728.000000          0.5000        0.5100   
...               ...        ...           ...             ...           ...   
5292         0.253333      198.0   2548.333333          0.1200        0.0875   
5304         0.360000      209.0   2969.000000          0.1150        0.0950   
5433              NaN      241.0   5545.000000          0.8700        0.9300   
5440              NaN      247.0   5564.000000          1.1475        1.0550   
5512         0.650000      250.0   3158.000000          0.1800        0.1700   

       distance spectral_type  
171   551.88600             0  
179     4.84867        M5.5 V  
180     4.84867        M5.5 V  
237     9.44181        M2.5 V  
263     7.61778    M0.5-1.0 V  
...         ...           ...  
5292        NaN             0  
5304    3.83078        M7.0 V  
5433  163.37100             0  
5440  264.78000             0  
5512    9.58341        M5.0 V  

[78 rows x 15 columns]

In [145]:
# counts after removal of controversies
proj_data2.count()

index                     5572
planet_name               5572
num_stars                 5572
orbital_period            5326
planet_semi-major_axis    5280
planet_radius             4231
planet_mass               2674
eccentricity              4804
insolation_flux           3885
equi_temp                 4130
stellar_temp              5184
stellar_radius            5049
stellar_mass              5549
distance                  5456
spectral_type             5572
dtype: int64

# Solving Missing Equilibrium Temperature

In [146]:
# Does insolation_flux help?
miss_temp = k_data[k_data['equi_temp'].isna()]
#miss_temp.count()
# 93 counts with insolation flux
insol_flux = k_data[~k_data['insolation_flux'].isna()]
insol_flux.head()

index  planet_name  num_stars  orbital_period  planet_semi-major_axis  \
30     30     55 Cnc e        2.0        0.994802                0.019644   
47     47     AU Mic b        1.0        8.463095                0.064860   
48     48     AU Mic c        1.0       18.859087                0.110250   
74     75  BD-210397 b        1.0     1891.000000                2.630000   
75     76  BD-210397 c        1.0     6300.000000                5.900000   

    planet_radius  planet_mass  eccentricity  insolation_flux    equi_temp  \
30       1.968556     8.667005      0.081625          2657.83  1894.500000   
47       4.089400    22.221567      0.064539            48.92   649.333333   
48       2.850667    14.900000      0.043595             7.30   456.500000   
74            NaN   214.000000      0.100000             0.02          NaN   
75            NaN  1142.288175      0.265000             0.00          NaN   

    stellar_temp  stellar_radius  stellar_mass  distance spectral_type  
30   5239.990833         0.95500      0.922667   12.5855          G8 V  
47   3675.857143         0.75125      0.515714    9.7221            M1  
48   3685.750000         0.76200      0.522000    9.7221             0  
74   4051.000000         0.75000      0.680000   23.7324          K7 V  
75   4051.000000         0.75000      0.685000   23.7324          K7 V

In [147]:
insol_habit = insol_flux[(insol_flux['insolation_flux'] >= 175) & (insol_flux['insolation_flux'] <= 300)]
both_habit = insol_habit[(insol_habit['equi_temp'] < 800)]
both_habit

index    planet_name  num_stars  orbital_period  planet_semi-major_axis  \
2343   2362  Kepler-1283 b        1.0       12.946239                 0.10696   

      planet_radius  planet_mass  eccentricity  insolation_flux  equi_temp  \
2343       1.706571          NaN           0.0          210.312      771.0   

      stellar_temp  stellar_radius  stellar_mass  distance spectral_type  
2343   6089.444444        1.498889       1.04125   2921.99             0

# Conclusion on Insolation Flux

Useless parameter, the planet's who's insolation flux have an equilibrium temperature that is too high to sustain life.

In [148]:
#removing insolation flux from dataframe
k_data = k_data.drop(columns=['insolation_flux'])
no_flux = k_data
no_flux.head()

index planet_name  num_stars  orbital_period  planet_semi-major_axis  \
0      0    11 Com b        2.0      324.620000                1.226000   
1      1    11 UMi b        1.0      516.219985                1.526667   
2      2    14 And b        1.0      186.300000                0.761667   
3      3    14 Her b        1.0     1766.378417                2.814750   
4      4  16 Cyg B b        3.0      799.375000                1.662833   

   planet_radius  planet_mass  eccentricity  equi_temp  stellar_temp  \
0            NaN  5505.066163      0.234500        NaN      4808.000   
1            NaN  3818.094733      0.080000        NaN      4276.500   
2            NaN  1224.550433      0.000000        NaN      4850.500   
3            NaN  1642.383591      0.362925        NaN      5296.985   
4            NaN   533.514528      0.676033        NaN      5728.594   

   stellar_radius  stellar_mass  distance spectral_type  
0       16.380000      2.463333   93.1846        G8 III  
1       26.935000      2.093333  125.3210        K4 III  
2       11.275000      1.726667   75.4392        K0 III  
3        0.976667      0.927143   17.9323          K0 V  
4        1.140000      1.016000   21.1397        G2.5 V

# Continuation of finding NAN equi_temp

Let $T_{s} = $ temperature of star, $R = $ radius of star, $a = $ semi-major axis, and $A_{B} = $ Bond Albedo

$T_{eq} = T_{s}\sqrt{\frac{R}{2a}}(1-A_{B})^{\frac{1}{4}}$

However, we don't have Bond Albedo, and if we did then the scientists that have been focused on this effort would have already done the math...

BUT WE CAN ESTIMATE!!!

Earth's Bond Albedo is 0.29

Bond Albedo is primarily affected by atmosphere. More clouds, more albedo.

Would it be wrong to take an average of all of our planet's bond albedos as our $A_{B}$?

In [149]:
no_flux.head()

index planet_name  num_stars  orbital_period  planet_semi-major_axis  \
0      0    11 Com b        2.0      324.620000                1.226000   
1      1    11 UMi b        1.0      516.219985                1.526667   
2      2    14 And b        1.0      186.300000                0.761667   
3      3    14 Her b        1.0     1766.378417                2.814750   
4      4  16 Cyg B b        3.0      799.375000                1.662833   

   planet_radius  planet_mass  eccentricity  equi_temp  stellar_temp  \
0            NaN  5505.066163      0.234500        NaN      4808.000   
1            NaN  3818.094733      0.080000        NaN      4276.500   
2            NaN  1224.550433      0.000000        NaN      4850.500   
3            NaN  1642.383591      0.362925        NaN      5296.985   
4            NaN   533.514528      0.676033        NaN      5728.594   

   stellar_radius  stellar_mass  distance spectral_type  
0       16.380000      2.463333   93.1846        G8 III  
1       26.935000      2.093333  125.3210        K4 III  
2       11.275000      1.726667   75.4392        K0 III  
3        0.976667      0.927143   17.9323          K0 V  
4        1.140000      1.016000   21.1397        G2.5 V

Using 11 Com b as a trial run. $0.3535 = $ average bond albedo sol system planets. Acknowledging this probably only works for G Type stars that are similar to Sol.

Let $T_{s} = 4808$  $R = 11395566$, $a = 183407176$, and $A_{B} = 0.3535$


$T_{eq} = (4808)\sqrt{\frac{11395566}{2 * 183407176}}(1 - 0.3535)^{\frac{1}{4}}$

$T_{eq} = (4808)\sqrt{\frac{11395566}{366814352}}(0.6465)^{\frac{1}{4}}$

$T_{eq} = 847.44(0.89669)$

$T_{eq} = 759.89$

Does this seem right? $1.22$ times further from star. The star is a giant but is colder than Sol. $16$ times larger radius really 

Let's try this same calculation with Earth and see if it's at least similar.

Let $T_{s} = 5800$  $R = 695700$, $a = 149598023$, and $A_{B} = 0.3535$


$T_{eq} = (5800)\sqrt{\frac{695700}{2 * 149598023}}(1 - 0.3535)^{\frac{1}{4}}$

$T_{eq} = 279.68(0.89669)$

$T_{eq} = 250.79$

Actual $= 255$

Error $= |Actual - Estimate| = 4.21$

While that is not how one is supposed to calculate inter-solar equilibrium temperature, that was closer than what I was expecting.

In [150]:
#Function to fill in missing equi_temp
def calc_equi_temp(row):
    if np.isnan(row['equi_temp']): 
        stellar_temp = row['stellar_temp']
        stellar_radius = row['stellar_radius']
        a = row['planet_semi-major_axis']
        stellar_rad = stellar_radius * 695700
        b = a * 149598023
        #print(stellar_temp)
        #print(stellar_rad)
        #print(b)
        return stellar_temp * np.sqrt(stellar_rad / (2 * b)) * 0.89669
    else:
        return row['equi_temp']

In [151]:
trial = no_flux.copy(deep=True)

In [152]:
trial['equi_temp'] = trial.apply(calc_equi_temp, axis=1)
trial.head()

index planet_name  num_stars  orbital_period  planet_semi-major_axis  \
0      0    11 Com b        2.0      324.620000                1.226000   
1      1    11 UMi b        1.0      516.219985                1.526667   
2      2    14 And b        1.0      186.300000                0.761667   
3      3    14 Her b        1.0     1766.378417                2.814750   
4      4  16 Cyg B b        3.0      799.375000                1.662833   

   planet_radius  planet_mass  eccentricity   equi_temp  stellar_temp  \
0            NaN  5505.066163      0.234500  759.891481      4808.000   
1            NaN  3818.094733      0.080000  776.694550      4276.500   
2            NaN  1224.550433      0.000000  806.933439      4850.500   
3            NaN  1642.383591      0.362925  134.914123      5296.985   
4            NaN   533.514528      0.676033  205.093318      5728.594   

   stellar_radius  stellar_mass  distance spectral_type  
0       16.380000      2.463333   93.1846        G8 III  
1       26.935000      2.093333  125.3210        K4 III  
2       11.275000      1.726667   75.4392        K0 III  
3        0.976667      0.927143   17.9323          K0 V  
4        1.140000      1.016000   21.1397        G2.5 V

In [153]:
#Remove the two White Dwarves, did the math, wouldn't matter to the data
trial = trial[~trial['spectral_type'].str.contains('WD')]

In [154]:
# Now let's see how many stars we have in the habitable.
habitable2 = trial[(trial['equi_temp'] >= 175) & (trial['equi_temp'] <= 270)]
count2 = habitable2['planet_name'].count()
missing2 = trial['equi_temp'].isna().sum()
print(count2, "planets with equilibrium temp within acceptable habitable zone parameters.", missing2, "NAN values")

212 planets with equilibrium temp within acceptable habitable zone parameters. 660 NAN values


In [155]:
still_missing = trial[trial['equi_temp'].isna()]
still_missing.count()

index                     660
planet_name               660
num_stars                 660
orbital_period            449
planet_semi-major_axis    476
planet_radius             140
planet_mass               524
eccentricity              302
equi_temp                   0
stellar_temp              288
stellar_radius            170
stellar_mass              642
distance                  647
spectral_type             660
dtype: int64

# Problem

If stellar temp, stellar radius, or planet_sma are NAN then unable to find equilibrium, meaning 600 possible planets are still unknown.

For stellar temp, since we know the spectral type in some instances we can further approximate. 

Will use this page for temperature. https://sites.uni.edu/morgans/astro/course/Notes/section2/spectraltemps.html

For Stellar Radius we may have to interpolate.

Rows without spectral types will be removed below.

In [156]:
#Removes rows missing spectral type. Lose ~400 rows
still_missing = still_missing[still_missing['spectral_type'] != '0']
still_missing.count()

index                     215
planet_name               215
num_stars                 215
orbital_period            192
planet_semi-major_axis    167
planet_radius              28
planet_mass               188
eccentricity              159
equi_temp                   0
stellar_temp              160
stellar_radius             43
stellar_mass              213
distance                  212
spectral_type             215
dtype: int64

In [157]:
#separating spectral_type into two
split_columns = still_missing['spectral_type'].str.split(expand=True)
still_missing['spectral_type'] = split_columns[0]
still_missing['spectral_subtype'] = split_columns[1] if len(split_columns.columns) > 1 else np.nan
still_missing.count()

index                     215
planet_name               215
num_stars                 215
orbital_period            192
planet_semi-major_axis    167
planet_radius              28
planet_mass               188
eccentricity              159
equi_temp                   0
stellar_temp              160
stellar_radius             43
stellar_mass              213
distance                  212
spectral_type             215
spectral_subtype          146
dtype: int64

In [158]:
# function to fill in missing stellar_temp
def fill_stellar_temp(df):
    III_mapping = {
        #O type Full Info
        'O5': 54000,
        'O6': 45000,
        'O7': 43300,
        'O8': 40600,
        'O9': 37800,
        
        #B type Full Info
        'B0': 29200,
        'B1': 23000,
        'B2': 21000,
        'B3': 17600,
        'B5': 15200,
        'B6': 14300,
        'B7': 13500,
        'B8': 12300,
        'B9': 11400,
        
        #A type Full Info
        'A0': 9600,
        'A1': 9330,
        'A2': 9040,
        'A3': 8750,
        'A4': 8480,
        'A5': 8310,
        'A7': 7920,
        
        #F type Full Info
        'F0': 7350,
        'F2': 7050,
        'F3': 6850,
        'F5': 6700,
        'F6': 6550,
        'F7': 6400,
        'F8': 6300,
        
        #G type Full Info
        'G0': 6050,
        'G1': 5930,
        'G2': 5800,
        'G5': 5660,
        'G8': 5440,
        
        #K type Full Info
        'K0': 5240,
        'K1': 5110,
        'K2': 4960,
        'K3': 4800,
        'K4': 4600,
        'K5': 4400,
        'K7': 4000,
        
        #M type Full Info
        'M0': 3750,
        'M1': 3700,
        'M2': 3600,
        'M3': 3500,
        'M4': 3400,
        'M5': 3200,
        'M6': 3100,
        'M7': 3900,
        'M8': 2700
    }
    
    for index, row in df.iterrows():
        if pd.isna(row['stellar_temp']):
            spectral_type = row['spectral_type']
            if spectral_type in III_mapping:
                df.at[index, 'stellar_temp'] = III_mapping[spectral_type]
    
    return df

filled_trial = fill_stellar_temp(still_missing)
filled_trial.head(50)

index                   planet_name  num_stars  orbital_period  \
12      12  2MASS J01033563-5515561 AB b        2.0             NaN   
15      15     2MASS J04414489+2301513 b        1.0             NaN   
18      18     2MASS J22362452+4751425 b        1.0             NaN   
45      45                      AB Pic b        1.0             NaN   
58      58                   BD+45 564 b        1.0      307.880000   
62      62                   BD+55 362 b        1.0      265.590000   
64      64                  BD+63 1405 b        1.0     1214.240000   
65      65                  BD-00 4475 b        1.0      723.200000   
66      67                  BD-06 1339 c        1.0      125.940000   
68      69                  BD-08 2823 c        1.0      237.600000   
79      80                     CHXR 73 b        1.0             NaN   
116    117                     CoRoT-7 d        1.0        8.966000   
121    122    DENIS-P J082303.1-491201 b        1.0      246.360000   
194    196                    GJ 160.2 b        1.0        5.235400   
199    201                      GJ 180 b        1.0       17.256500   
200    202                      GJ 180 c        1.0       24.329000   
206    208                      GJ 229 b        2.0      525.529983   
208    210                     GJ 27.1 b        1.0       15.819000   
238    240                     GJ 3634 b        1.0        2.645610   
240    242                      GJ 367 c        1.0       11.530100   
241    243                      GJ 367 d        1.0       34.369000   
253    255                      GJ 422 b        1.0       23.145000   
256    258                      GJ 433 c        1.0     4247.052500   
260    262                      GJ 480 b        1.0        9.567000   
270    272                    GJ 667 C b        3.0        7.200792   
271    273                    GJ 667 C c        3.0       28.134500   
272    277                      GJ 674 b        1.0        4.693800   
275    280                    GJ 676 A d        2.0        3.600420   
276    281                    GJ 676 A e        2.0       35.411193   
278    283                      GJ 682 b        1.0       17.478000   
279    284                      GJ 682 c        1.0       57.320000   
298    303                    GJ 896 A b        2.0      284.390000   
301    306                       GJ 96 b        1.0       73.940000   
477    483                   HD 102843 b        1.0     3090.942000   
481    487                   HD 103774 b        2.0        5.888100   
483    489                   HD 103949 b        1.0      120.878000   
501    507                   HD 108202 b        1.0     2990.000000   
512    518                   HD 109271 b        1.0        7.854300   
513    519                   HD 109271 c        1.0       30.930000   
533    539                   HD 111232 c        1.0    29286.223115   
572    578                   HD 124330 b        1.0      270.660000   
573    579                    HD 12484 b        1.0       58.830000   
589    595                   HD 129445 b        1.0     1821.000000   
596    602                 HD 133131 A b        2.0      649.000000   
597    603                 HD 133131 A c        2.0     3568.000000   
598    604                 HD 133131 B b        2.0     5769.000000   
599    605                   HD 134060 b        1.0        3.269600   
600    606                   HD 134060 c        1.0     1291.564600   
608    614                   HD 135625 b        1.0     4055.000000   
610    616                   HD 136118 b        1.0     1189.298825   

     planet_semi-major_axis  planet_radius  planet_mass  eccentricity  \
12                84.000000            NaN  4131.790000           NaN   
15                15.000000            NaN  2383.600000           NaN   
18               230.000000            NaN  3972.875000           NaN   
45               260.000000            NaN  4290.500000           NaN   
58                

In [159]:
#nan_count = filled_trial['stellar_temp'].isna().sum()
#nan_count

In [160]:
#filled_trial[filled_trial['stellar_temp'] == 4960].count()

In [161]:
#trial[trial['stellar_temp'] == 4960].count()

In [162]:
filled_trial = filled_trial.dropna(subset=['stellar_temp'])

In [163]:
filled_trial.count()

index                     187
planet_name               187
num_stars                 187
orbital_period            179
planet_semi-major_axis    139
planet_radius              28
planet_mass               160
eccentricity              147
equi_temp                   0
stellar_temp              187
stellar_radius             42
stellar_mass              185
distance                  187
spectral_type             187
spectral_subtype          132
dtype: int64

In [164]:
# Okay we now have 187 additional stellar temps calculated. 
# Let's fill in missing Semi-Major_axis based upon orbital period
filled_trial.loc[filled_trial['planet_semi-major_axis'].isna(), 'planet_semi-major_axis'] = (filled_trial['orbital_period']**2)**(1/3)
filled_trial.count()
#filled_trial.head(50)

index                     187
planet_name               187
num_stars                 187
orbital_period            179
planet_semi-major_axis    187
planet_radius              28
planet_mass               160
eccentricity              147
equi_temp                   0
stellar_temp              187
stellar_radius             42
stellar_mass              185
distance                  187
spectral_type             187
spectral_subtype          132
dtype: int64

In [165]:
# Now we see we filled in missing planet_sma based upon Kepler's Third Law.
# Once more we can run our calc_equi_temp on 42 exo-planets since we have 42 stellar radius
# However that leaves us with 145 planets still unable to calculate due to missing stellar radius.
filled_trial['equi_temp'] = filled_trial.apply(calc_equi_temp, axis=1)
filled_trial.count()

/tmp/ipykernel_130/3416976009.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  filled_trial['equi_temp'] = filled_trial.apply(calc_equi_temp, axis=1)


index                     187
planet_name               187
num_stars                 187
orbital_period            179
planet_semi-major_axis    187
planet_radius              28
planet_mass               160
eccentricity              147
equi_temp                  42
stellar_temp              187
stellar_radius             42
stellar_mass              185
distance                  187
spectral_type             187
spectral_subtype          132
dtype: int64

In [166]:
habitable3 = filled_trial[(filled_trial['equi_temp'] >= 175) & (filled_trial['equi_temp'] <= 270)]
count3 = habitable3['planet_name'].count()
missing3 = filled_trial['equi_temp'].isna().sum()
print(count3, "planets with equilibrium temp within acceptable habitable zone parameters.", missing3, "NAN values")

3 planets with equilibrium temp within acceptable habitable zone parameters. 145 NAN values


In [167]:
# Merge three habitable dataframes

habitable_true = pd.concat([habitable, habitable2, habitable3], ignore_index=True)
habitable_true.count()

index                     293
planet_name               293
num_stars                 293
orbital_period            293
planet_semi-major_axis    289
planet_radius             130
planet_mass               197
eccentricity              270
insolation_flux            70
equi_temp                 293
stellar_temp              293
stellar_radius            291
stellar_mass              293
distance                  275
spectral_type             293
spectral_subtype            0
dtype: int64

In [168]:
# Secondary clean up due to messy data being merged.

# Remove Flux once again
habitable_true.drop(columns=['insolation_flux'])

# Split spectral type
split_columns = habitable_true['spectral_type'].str.split(expand=True)
habitable_true['spectral_type'] = split_columns[0]
habitable_true['spectral_subtype'] = split_columns[1] if len(split_columns.columns) > 1 else np.nan

In [174]:
# Something is wrong. Habit1 = 78, Habit2 = 212, Habit3 = 3
# Found it. Accidentally ran concat more than once. 
# But also found out each habitable had duplicates that needed to be removed.
habitable_true.drop_duplicates(subset='planet_name', inplace=True)

habitable_true.reset_index(inplace=True)
habitable_true

level_0  index       planet_name  num_stars  orbital_period  \
0          0    172  EPIC 248847494 b        1.0     3650.000000   
1          1    180         GJ 1002 b        1.0       10.346500   
2          2    181         GJ 1002 c        1.0       21.202000   
3          3    239          GJ 357 d        1.0       55.661000   
4          4    265          GJ 514 b        1.0      140.430000   
..       ...    ...               ...        ...             ...   
210      277   5091         TOI-199 c        1.0      273.690000   
211      289   5599         ups And d        2.0     1285.346167   
212      290    707       HD 158259 c        1.0        3.432000   
213      291   1717          K2-396 b        1.0        0.673862   
214      292   1734          K2-409 b        1.0        1.908084   

     planet_semi-major_axis  planet_radius  planet_mass  eccentricity  \
0                  4.500000         12.442   4131.79000      0.000000   
1                  0.045700            NaN      1.08000           NaN   
2                  0.073800            NaN      1.36000           NaN   
3                  0.204000            NaN      6.10000           NaN   
4                  0.422000            NaN      5.20000      0.450000   
..                      ...            ...          ...           ...   
210                0.807000            NaN     88.99195      0.096000   
211                2.528382            NaN   1616.24259      0.281117   
212                2.275263            NaN      5.60000           NaN   
213                0.768624          1.579          NaN           NaN   
214                1.538385          2.482          NaN           NaN   

     insolation_flux   equi_temp  stellar_temp  stellar_radius  stellar_mass  \
0               0.19  183.000000   4898.000000           2.700        0.9000   
1               0.67  231.000000   3024.000000           0.140        0.1200   
2               0.26  182.000000   3024.000000           0.140        0.1200   
3               0.38  220.000000   3505.000000           0.340        0.3400   
4               0.28  202.000000   3728.000000           0.500        0.5100   
..               ...         ...           ...             ...           ...   
210              NaN  229.043664   5255.000000           0.820        0.9400   
211              NaN  211.390496   6117.093333           1.615        1.2625   
212              NaN  190.768956   6050.000000           1.210        1.0800   
213              NaN  247.691992   5236.000000           0.920        0.9500   
214              NaN  222.002901   5742.000000           1.230        0.8900   

      distance spectral_type spectral_subtype  
0    551.88600             0             None  
1      4.84867          M5.5                V  
2      4.84867          M5.5                V  
3      9.44181          M2.5                V  
4      7.61778      M0.5-1.0                V  
..         ...           ...              ...  
210  102.27000            G9                V  
211   13.40540            F8                V  
212   27.04980            G0             None  
213  274.86400            G9             None  
214  431.03100            G4             None  

[215 rows x 17 columns]